File: eurorack_case/params.py
"""Parameters for the asymmetric Eurorack case.

Names follow the variables in the Onshape source (reference/Eurorackv1asym.fs)
so the two can be read side by side. The first block is the Onshape
configuration (what the user sets); the rest are the design constants that
were `#variables` in the Part Studio.

Units are millimetres throughout.
"""
from dataclasses import dataclass

HP = 5.08

LIP_STYLES = ("round", "sharp", "chamfer")


@dataclass(frozen=True)
class CaseParams:
    # --- configuration inputs -------------------------------------------
    hp_count: int = 26
    front_height: float = 30.0      # Front_Height: outer height of the front wall
    rear_height: float = 30.0       # RearHeight:   outer height of the rear wall
    top_lips: str = "round"         # round | sharp | chamfer
    upper_round: float = 1.0
    upper_chamfer: float = 1.0
    bottom_lips: str = "round"
    lower_round: float = 1.0
    lower_chamfer: float = 1.0      # Lower_Lip_Bevel
    left_wall: bool = False         # True = the "asym" Part Studio: X = 0 closed by an
                                    # integral wall, one removable cap at the far end.
                                    # False = the "sym" Part Studio: open both ends, two caps.

    # --- shell ----------------------------------------------------------
    panel_height: float = 129.5     # PanelHeight: clear width between the inner walls
    panel_thickness: float = 2.0    # PanelThickness: depth of the panel recess
    wall_thickness: float = 4.0     # CaseFrontBackThickness
    bottom_thickness: float = 4.0   # CaseBottomThickness

    # --- rails (T-slot blocks under the panel recess) -------------------
    rail_depth: float = 7.798472    # drawn with a bare 7.798 in the sketch; the RailBlockDepth
                                    # variable (8 mm) is never referenced. Matches the STEP.
    rail_height: float = 10.0       # solved height of the rail block
    rail_inset: float = 3.0         # RailBlockPanelInset: bolt axis from the inner wall
    rail_bolt_width: float = 3.2    # RailBlockBoltWidth
    rail_bolt_depth: float = 8.0    # RailBlockBoltDepth
    rail_nut_width: float = 5.8     # RailBlockNutWidth: sizes the nut traps in the end cap tab
    rail_nut_slot_width: float = 5.596945   # the rail's sliding-nut slot is undimensioned in
                                    # the sketch and solved to this. Matches the STEP.
    rail_nut_thickness: float = 2.5 # RailBlockNutThickness
    rail_nut_depth: float = 2.0     # RailBlockNutDepth: nut pocket top below the rail top
    nut_clearance: float = 0.1      # pockets are cut RailBlockNutWidth - .1mm

    # --- end caps -------------------------------------------------------
    end_cap_thickness: float = 4.0  # EndCapThickness (= wall thickness in the source)
    end_cap_tab_thickness: float = 5.0   # EndCapTabThickness
    end_cap_face_chamfer: float = 1.5    # edge break around the outer face
    tab_clearance: float = 0.0      # shrink the tab for a printable fit (0 in the source)

    # --- end cap screws (M3, through the shell into the tab) ------------
    bolt_through_diam: float = 3.3      # BoltThrougDiam
    bolt_counterbore_diam: float = 6.0  # BoltCounterboreDiam
    bolt_counterbore_depth: float = 1.75  # BoltCounterboreThick
    bottom_hole_inset: float = 4.0      # BottomHoleInset: screw axis from the case end
    bottom_hole_spacing: float = 55.0   # BottomHoleSpacing
    wall_hole_spacing: float = 23.0     # RearHoleSpacing

    # --- blank test panel ----------------------------------------------
    panel_hole_diam: float = 3.2
    panel_hole_edge_inset: float = 3.0  # PanelHoleVerticalInset
    panel_hole_side_inset: float = 7.5

    def __post_init__(self):
        if self.top_lips not in LIP_STYLES or self.bottom_lips not in LIP_STYLES:
            raise ValueError(f"lip style must be one of {LIP_STYLES}")
        if self.hp_count < 2:
            raise ValueError("hp_count must be at least 2")
        minimum = self.panel_thickness + self.rail_height + self.bottom_thickness + self.end_cap_tab_thickness
        for name in ("front_height", "rear_height"):
            if getattr(self, name) < minimum:
                raise ValueError(
                    f"{name} must be at least {minimum} mm to fit the end cap tab between the rail and the floor"
                )

    # --- derived --------------------------------------------------------
    @property
    def width(self) -> float:
        """Case length along the rails."""
        return self.hp_count * HP

    @property
    def inner_half(self) -> float:
        return self.panel_height / 2

    @property
    def outer_half(self) -> float:
        return self.inner_half + self.wall_thickness

    @property
    def rail_top(self) -> float:
        return -self.panel_thickness

    @property
    def rail_bottom(self) -> float:
        return -self.panel_thickness - self.rail_height

    @property
    def nut_pocket_width(self) -> float:
        """Square nut traps in the end cap tab: RailBlockNutWidth - .1mm."""
        return self.rail_nut_width - self.nut_clearance

    @property
    def closed_ends(self):
        return ("left",) if self.left_wall else ()

    @property
    def end_cap_tab_depth(self) -> float:
        """EndCapTabDepth = #BottomHoleInset + .7 * #BoltCounterboreDiam."""
        return self.bottom_hole_inset + 0.7 * self.bolt_counterbore_diam

File: web/src/model/params.ts
/**
 * Parameters for the Eurorack case — a port of eurorack_case/params.py.
 *
 * Names follow the Onshape variable table (docs/variables.md). The first block
 * is the Onshape configuration; the rest are the design constants that were
 * `#variables` in the Part Studio. Millimetres throughout.
 */

export const HP = 5.08;

export type LipStyle = "round" | "sharp" | "chamfer";
export const LIP_STYLES: readonly LipStyle[] = ["round", "sharp", "chamfer"];

export interface CaseParamsInput {
  // --- configuration inputs ---------------------------------------------
  hpCount?: number;
  frontHeight?: number;        // Front_Height: outer height of the front wall
  rearHeight?: number;         // RearHeight:   outer height of the rear wall
  topLips?: LipStyle;
  upperRound?: number;
  upperChamfer?: number;
  bottomLips?: LipStyle;
  lowerRound?: number;
  lowerChamfer?: number;       // Lower_Lip_Bevel
  leftWall?: boolean;          // true = "asym" Part Studio: X = 0 closed by an integral wall

  // --- shell --------------------------------------------------------------
  panelHeight?: number;        // PanelHeight: clear width between the inner walls
  panelThickness?: number;     // PanelThickness: depth of the panel recess
  wallThickness?: number;      // CaseFrontBackThickness
  bottomThickness?: number;    // CaseBottomThickness

  // --- rails ----------------------------------------------------------------
  railDepth?: number;          // drawn 7.798 in the sketch; RailBlockDepth (8) is never referenced
  railHeight?: number;         // solved height of the rail block
  railInset?: number;          // RailBlockPanelInset: bolt axis from the inner wall
  railBoltWidth?: number;      // RailBlockBoltWidth
  railBoltDepth?: number;      // RailBlockBoltDepth
  railNutWidth?: number;       // RailBlockNutWidth: sizes the nut traps in the tab
  railNutSlotWidth?: number;   // the rail's sliding-nut slot: undimensioned, solved to 5.597
  railNutThickness?: number;   // RailBlockNutThickness
  railNutDepth?: number;       // RailBlockNutDepth
  nutClearance?: number;       // traps are cut RailBlockNutWidth - .1mm

  // --- end caps ---------------------------------------------------------------
  endCapThickness?: number;    // EndCapThickness (= wall thickness in the source)
  endCapTabThickness?: number; // EndCapTabThickness
  endCapFaceChamfer?: number;  // edge break around the cap's outer face
  tabClearance?: number;       // shrink the tab for a printable fit (0 in the source)

  // --- end cap screws (M3, through the shell into the tab) -------------------
  boltThroughDiam?: number;    // BoltThrougDiam
  boltCounterboreDiam?: number;  // BoltCounterboreDiam
  boltCounterboreDepth?: number; // BoltCounterboreThick
  bottomHoleInset?: number;    // BottomHoleInset: screw axis from the case end
  bottomHoleSpacing?: number;  // BottomHoleSpacing
  wallHoleSpacing?: number;    // RearHoleSpacing

  // --- blank test panel -------------------------------------------------------
  panelHoleDiam?: number;
  panelHoleEdgeInset?: number; // PanelHoleVerticalInset
  panelHoleSideInset?: number;
}

export type CaseParams = Required<CaseParamsInput> & {
  readonly width: number;
  readonly innerHalf: number;
  readonly outerHalf: number;
  readonly outerDepth: number;      // Onshape's measured `BaseCase`, 137.5
  readonly railTop: number;
  readonly railBottom: number;
  readonly nutPocketWidth: number;
  readonly endCapTabDepth: number;  // #BottomHoleInset + .7 * #BoltCounterboreDiam
  readonly closedEnds: readonly ("left" | "right")[];
};

export const DEFAULTS: Required<CaseParamsInput> = {
  hpCount: 26,
  frontHeight: 30,
  rearHeight: 30,
  topLips: "round",
  upperRound: 1,
  upperChamfer: 1,
  bottomLips: "round",
  lowerRound: 1,
  lowerChamfer: 1,
  leftWall: false,

  panelHeight: 129.5,
  panelThickness: 2,
  wallThickness: 4,
  bottomThickness: 4,

  railDepth: 7.798472,
  railHeight: 10,
  railInset: 3,
  railBoltWidth: 3.2,
  railBoltDepth: 8,
  railNutWidth: 5.8,
  railNutSlotWidth: 5.596945,
  railNutThickness: 2.5,
  railNutDepth: 2,
  nutClearance: 0.1,

  endCapThickness: 4,
  endCapTabThickness: 5,
  endCapFaceChamfer: 1.5,
  tabClearance: 0,

  boltThroughDiam: 3.3,
  boltCounterboreDiam: 6,
  boltCounterboreDepth: 1.75,
  bottomHoleInset: 4,
  bottomHoleSpacing: 55,
  wallHoleSpacing: 23,

  panelHoleDiam: 3.2,
  panelHoleEdgeInset: 3,
  panelHoleSideInset: 7.5,
};

export function caseParams(input: CaseParamsInput = {}): CaseParams {
  const p = { ...DEFAULTS, ...input };
  if (!LIP_STYLES.includes(p.topLips) || !LIP_STYLES.includes(p.bottomLips)) {
    throw new Error(`lip style must be one of ${LIP_STYLES.join(", ")}`);
  }
  if (!Number.isInteger(p.hpCount) || p.hpCount < 2) throw new Error("hpCount must be an integer of at least 2");
  const minimum = p.panelThickness + p.railHeight + p.bottomThickness + p.endCapTabThickness;
  for (const name of ["frontHeight", "rearHeight"] as const) {
    if (p[name] < minimum) {
      throw new Error(`${name} must be at least ${minimum} mm to fit the end cap tab between the rail and the floor`);
    }
  }
  // the wall top is only wallThickness wide before it drops into the panel recess,
  // so an upper lip has to be smaller than that to have a face to land on
  const upper = p.topLips === "round" ? p.upperRound : p.topLips === "chamfer" ? p.upperChamfer : 0;
  if (upper >= p.wallThickness) {
    throw new Error(`the top lip (${upper} mm) must be smaller than the wall thickness (${p.wallThickness} mm)`);
  }
  const innerHalf = p.panelHeight / 2;
  const outerHalf = innerHalf + p.wallThickness;
  return {
    ...p,
    width: p.hpCount * HP,
    innerHalf,
    outerHalf,
    outerDepth: 2 * outerHalf,
    railTop: -p.panelThickness,
    railBottom: -p.panelThickness - p.railHeight,
    nutPocketWidth: p.railNutWidth - p.nutClearance,
    endCapTabDepth: p.bottomHoleInset + 0.7 * p.boltCounterboreDiam,
    closedEnds: p.leftWall ? ["left"] : [],
  };
}
